Enable bootstrap support

huddlej opened the pull request:
## Description of proposed changes

This PR enables annotation of bootstrap support from IQ-TREE (at least) in the raw, refined, and Auspice JSON trees through the following changes to Augur logic:

1. Allow users to override default tree builder arguments with `--tree-builder-args` values by toggling a new `--override-default-args` flag. This change maintains the original behavior of the tree builder args argument which is to augment the default arguments. **This change is backward compatible**.
2. Save "confidence" annotations on trees from bootstrap analyses in the `augur refine` node data JSON output. **This change is a backward compatible**.

With these changes, I can run IQ-TREE's bootstrap analysis for a SARS-CoV-2 tree (as shown in the testing section below) and get an Auspice color-by of `confidence` that looks like this:

<img width="986" alt="Example confidence values for internal nodes of a test SARS-CoV-2 tree using IQ-TREE's ultrafast bootstrap mode" src="https://user-images.githubusercontent.com/85372/152584935-e49944a2-7236-4617-af52-bd00d37ddf34.png">

As part of these major changes, this PR also improves the user interface of `augur tree` by alerting users when they specify tree builder arguments that conflict with hardcoded defaults and providing more detailed examples of how to specify tree builder arguments in the help text.

## Related issue(s)

Fixes #666 
Fixes #778
Fixes #780

## Testing

 - [x] Tested by CI
 - [x] Tested with SARS-CoV-2 trees built with IQ-TREE and ultrafast bootstrap mode with the Nextstrain CI data for the ncov workflow:

```bash
augur tree \
  --alignment results/europe/filtered.fasta \
  --tree-builder-args '-bb 1000 -bnni -czb' \
  --override-default-args \
  --exclude-sites defaults/sites_ignored_for_tree_topology.txt              \
  --output results/europe/tree_raw.nwk \
  --nthreads 4
```

## Release plan

This PR is backward compatible and can be released as a feature release.

jameshadfield commented:
Looks good @huddlej - thanks. 

The bootstrap values make sense when we don't resolve polytomies in `augur refine` but are confusing when we do. For example, the internal nodes in the following screenshot do not have confidences because they are nodes added by treetime to resolve polytomies. (Addressing this, or even adding docs, might be beyond the scope of this PR.)
![image](https://user-images.githubusercontent.com/8350992/152080097-6e6bd04c-e663-4315-ab32-e434ca70f24e.png)

huddlej commented:
Thanks, @jameshadfield! I agree that there's an inherent flaw to this approach of mapping values from one tree's topology to a different downstream tree's topology. Some solutions could include:

 - using the original Newick tree as the divergence tree in the Auspice JSON
 - calculating confidence values for new (or all) nodes in TreeTime
 - forcing TreeTime to keep polytomies when confidence values are available

I wouldn't advocate for any of these, though.

Instead, I wonder if these internal nodes with missing data for confidence could end up being useful as a way of identifying nodes added by TreeTime, as long as we clearly document their meaning.

huddlej commented:
@sidneymbell Since you opened issue #666, I'm curious about your thoughts on this implementation, too.

sidneymbell commented:
This is wondeful, thank you so much @huddlej!

ammaraziz commented:
Hi @huddlej, 

Thank you for this.

Regarding our previous discussion on arguments such as `-asr`, I recommend adding a more explicit message in the `-h` output indicating that the recommended cli input for `--tree-builder-args` should be with the assignment operator. 

Current help says:

> extra arguments to be passed directly to the executable of the requested method (e.g., --tree-builder-args="-czb"). 

Instead maybe something along the lines of:

>  extra arguments to be passed directly to the executable of the requested method. Use assignment operator (e.g., --tree-builder-args="-czb") to avoid unexpected errors.

Additionally, in https://github.com/nextstrain/augur/issues/778 you mentioned that `augur tree -h` might output a command template. I think this is a great idea, both as it helps the user enter the command and because it further illustrates the assignment operator usage for `--tree-builder-args`. Here's my suggestion:

    augur tree -s <alignment> -m <substitution-model> -nt <nthreads> <tree-builder-args>="extra_args"

Either with the top line comment (`Build a tree using a variety of methods.`) or using `epilog=` in argparse.

Thanks!

ammaraziz commented on augur/tree.py:
```diff
@@ -348,8 +356,11 @@ def register_arguments(parser):
     parser.add_argument('--exclude-sites', type=str, help='file name of one-based sites to exclude for raw tree building (BED format in .bed files, second column in tab-delimited files, or one position per line)')
-    parser.add_argument('--tree-builder-args', type=str, default='', help='extra arguments to be passed directly to the executable of the requested tree method (e.g., --tree-builder-args="-czb")')
-
+    parser.add_argument('--tree-builder-args', type=str, help=f"""arguments to pass to the tree builder overriding the default arguments except for input alignment path, number of threads, and substitution model (e.g., --tree-builder-args="-czb" for IQ-TREE).
```
I recommend adding a line about assignment operator usage for `--tree-builder-args`, eg:

    Use assignment operator (e.g., --tree-builder-args="-czb") to avoid unexpected errors.

huddlej replied:
Good call. I added this line to the help.

ammaraziz commented:
@huddlej I am happy to test/review this PR if it's acceptable. No problem if there's a protocol in place for reviewers.

huddlej commented:
@ammaraziz, thank you for the review! I've updated the help text in 70b177c based on your recommended. Let me know if you have any comments about that implementation, though.

@jameshadfield After we chatted today, I added the new constraint that you cannot resolve polytomies in `augur refine` when your input tree has confidence values (6266328). I also added the `--override-default-args` flag, to make the new behavior in this PR optional (fd688f4). I'm actually happy with this implementation, as it does not add much complexity to the internals or the UI, but it provides the functionality we need and does not break our existing workflows.

@rneher This may be better discussed synchronously, but [James](https://github.com/nextstrain/augur/pull/839#issuecomment-1027495548) and [I](https://github.com/nextstrain/augur/pull/839#issuecomment-1028415714) have been discussing how `augur refine` should behave when the input tree has confidence values on internal nodes from bootstrap analyses. The solution we thought seemed most technically parsimonious and biologically correct is implemented in 6266328, but do you have concerns about this approach or other ideas how to handle this situation?

ammaraziz commented:
@huddlej Two thumbs up from me! Thanks again for driving this, looking forward to the change as it will make our pipeline much better.

huddlej commented:
@rneher notes:

> I would like to continue to be able to ignore bootstrap values in a tree and still resolve polytomies.

Based on this feedback, I've dropped the hard constraint that bootstrap values require the `--keep-polytomies` flag for `augur refine`. Without this breaking change, I don't think the remaining bits of this PR are too controversial.

@jameshadfield and I could discuss how to best document the need to use `--keep-polytomies`, for users who use bootstrap values and do not want the nodes with missing information.

trvrb commented:
This approach seems reasonable to me. I know bootstrap values have been a common request. Thanks for getting this implemented.
